pkg/matrix-diff/matrix-diff.go

```diff
@@ -0,0 +1,75 @@
+package matrixdiff
+
+import (
+	"fmt"
+
+	"github.com/openshift-kni/commatrix/pkg/types"
+)
+
+// Describes the status of an endpointslice.
+type epsStatus int
+
+// Used = the port is open on the node.
+// NotUsed = the port is not open on the node.
+// Missing = no endpointslice defined for an open port.
+const (
+	used epsStatus = iota
+	notUsed
+	missing
+)
+
+type MatrixDiff struct {
+	matrix        []types.ComDetails
+	cdToEPSStatus map[string]epsStatus
+}
+
+// Generates the diff between mat1 to mat2.
+func Generate(mat1 *types.ComMatrix, mat2 *types.ComMatrix) (MatrixDiff, error) {
+	cds := []types.ComDetails{}
+	epsStatus := map[string]epsStatus{}
+
+	for _, cd := range mat1.Matrix {
+		cds = append(cds, cd)
+		epsStatus[cd.String()] = used
+
+		if !mat2.Contains(cd) {
+			epsStatus[cd.String()] = notUsed
+		}
+	}
+
+	for _, cd := range mat2.Matrix {
+		cds = append(cds, cd)
+		epsStatus[cd.String()] = used
+
+		if !mat1.Contains(cd) {
+			epsStatus[cd.String()] = missing
+		}
+	}
+
+	cds = types.SortAndRemoveDuplicates(cds)
+
+	return MatrixDiff{cds, epsStatus}, nil
+}
+
+func (m *MatrixDiff) String() (string, error) {
+	colNames, err := types.GetComMatrixHeadersByFormat(types.FormatCSV)
+	if err != nil {
+		return "", fmt.Errorf("error getting commatrix CSV tags: %v", err)
+	}
+	diff := colNames + "\n"
+
+	for _, cd := range m.matrix {
+		switch m.cdToEPSStatus[cd.String()] {
+		case used:
+			diff += fmt.Sprintf("%s\n", cd)
+		case notUsed:
+			// add "+" before cds that have an endpointslice but the port is not open on the node
```

Comment:
its need to be general not diff from endpointslice?
what if we have 2 martix and we want to show diff maybe from 2 different  cluster?

Reply:
Matrix represents endpointslices in the cluster, what do you mean by general?

Reply:
like diff matrix is always between endpointslice matrix?

Reply:
We also do diff between the documented matrix and the endpoint slice matrix, so maybe the comment is not accurate.